src/lib/pickles/compile.ml

```diff
@@ -243,6 +248,37 @@ module Storables = struct
     }
 end
 
+let create_lock () =
+  let lock = ref (Promise.return ()) in
+
+  let open Promise.Let_syntax in
+  let run_in_sequence (f : unit -> 'a Promise.t) : 'a Promise.t =
+    (* acquire the lock *)
+    let existing_lock = !lock in
+    let unlock = ref (fun () -> ()) in
+    lock := Promise.create (fun resolve -> unlock := resolve) ;
+    (* await the existing lock *)
+    let%bind () = existing_lock in
+    (* run the function and release the lock *)
+    try
+      let%map res = f () in
+      !unlock () ; res
+    with exn -> !unlock () ; raise exn
+  in
+  run_in_sequence
```

Comment:
I don't quite get why this is useful: it's used only once below. What's the locking mechanism locking for?

Reply:
To not run circuits concurrently/ in an interlocking way, because that messes up snarky's global state

Reply:
It's used in a loop that runs many circuits btw